File: JavaScript/12_BinaryTree/LC_1120_MaximumAverageSubtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maximumAverageSubtree = function(root) {

    var max = 0;

    var dfs = function(root){
        if(!root){
            return [0.0,0.0];
        }

        let L = dfs(root.left);
        let R = dfs(root.right);

        var res = [L[0] + R[0] + 1.0,L[1] + R[1] + root.val];

        max = Math.max(max, res[1] / res[0]);

        return res;
    }

    dfs(root);
    return max;
};

File: Java/12_BinaryTree/LC_1120_MaximumAverageSubtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution_1120 {
    double max = Integer.MIN_VALUE;
    public double maximumAverageSubtree(TreeNode root) {
        dfs(root);
        return max;
    }

    private double[] dfs(TreeNode root){
        if(root == null){
            return new double[]{0.0,0.0};
        }

        double[] L = dfs(root.left);
        double[] R = dfs(root.right);

        double[] res = new double[2];
        res[0] = L[0] + R[0] + 1.0;
        res[1] = L[1] + R[1] + root.val;

        double averge = res[1] / res[0];
        max = Math.max(max,averge);

        return res;
    }
}